### PR TITLE
Cache the packageName to which receiver belongs

### PR DESCRIPTION
**Pull request by r30shah**
Cache the packageName to which receiver belongs in java/lang/Class.

Signed-off-by: Rahil Shah <rahil@ca.ibm.com>

### CODE REVIEW

**Comment by r30shah** [comment]
@tajila / @gacholio  Can I please get your review on this change? 
FYI @joransiu /  @vijaysun-omr

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 1]
```diff
@@ -2153,6 +2152,33 @@ String getPackageName() {
...
+ *
+ * @see			#getPackage
+ */
+/*[IF Sidecar19-SE]*/
+public
+/*[ENDIF] Sidecar19-SE */
+String getPackageName() {
+	String packageName = packageNameString;
+	if (null == packageName) {
+		synchronized (this) {
```
Comment (tajila) [inline jcl/src/java.base/share/classes/java/lang/Class.java:2171]: Why is this synchronized? If there is a race condition the result is always the same

**Comment by DanHeidinga** [comment]
@r30shah Can you also describe why this field is worth caching?  I'm assuming it's for performance reasons but it would help to identify the benchmarks where this is a benefit

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 1 above)
Comment (r30shah) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: Yes, we do not need this to be synchronized, I have made the change in https://github.com/eclipse-openj9/openj9/compare/32da8a53a24fc33450f4215c896c382ca89f5cac..f34d5c7b9557fb792d9ae874a3de46bf15342526

**Comment by r30shah** [comment]
@DanHeidinga , yes, we were looking into the cases where Method.invoke [1] calls in JDK11 lands in `slowVerifyAccess` [2] due to missing single level caller/receiver cache [3] we keep on AccessibleObject. While looking into the code-path where we spend significant amount of time in such case, I found that we call the `getPackageName` on Class object and compute the name of the package everytime. In such case, cached package name will be benefitial.
@joransiu  has prepared a simple benchmark using JMH framework to simulate such scenarios, which I am uploading here. [jmh-benchmark-joran.zip](https://github.com/eclipse-openj9/openj9/files/7549215/jmh-benchmark-joran.zip)
. From the set of benchmarks, microbenchmark `ReflectMethodInvokeBenchmark.testDifferentCallersInvokeStaticMethod` simulates such scenario where multiple classes calls a static method using same Method object forcing us to miss the single level caller/receiver cache on Accessible Object. Result is in the following table comparing baseline with the chage to cache the packageName.


| Benchmark | Baseline Throughput Performance | Cache PackageName Throughput Performance |
|-----------|---------------------------------|------------------------------------------|
| ReflectMethodInvokeBenchmark.testDifferentCallersInvokeStaticMethod | 100 | 415.63 |

[1]. https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/84f7b67e27b02ad1c8fe6f5d230fc28fc9025a39/src/java.base/share/classes/java/lang/reflect/Method.java#L549-L567
[2]. https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/84f7b67e27b02ad1c8fe6f5d230fc28fc9025a39/src/java.base/share/classes/java/lang/reflect/AccessibleObject.java#L630
[3]. https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/84f7b67e27b02ad1c8fe6f5d230fc28fc9025a39/src/java.base/share/classes/java/lang/reflect/AccessibleObject.java#L611-L616

**Comment by DanHeidinga** [comment]
@r30shah thanks for the JMH benchmark.  Is there any indication this is hitting real world workloads or other benchmarks?  There's a footprint / throughput tradeoff happening here and it would be good to know how prevalent this pattern is in the wild.

As an additional change, it probably makes sense to `intern()` the packageName String so there is only 1 copy in memory rather than 1 per class from the package.

**Comment by joransiu** [comment]
@DanHeidinga : This performance opportunity arose from our analysis of a customer reported issue - highlighted as a Java 11 performance difference vs Java 8 in their application.   The main cause of difference is due to Java 11's removal of `Reflection.quickCheckMemberAccess(clazz, modifiers)` from Method.invoke(), which had guarded calls to `getCallerClass()` and `checkAccess()`.  Now, under certain conditions, we end up stressing these previously guarded paths more often.

For this customer app, the main calls to Method.invoke that hits the slow verify path are from their application code.   However, the next hottest callers of Method.invoke are from from: `java.security.Provider.newInstanceUtil` driven by various call stacks related to Jar verifier and Log4J code.   The caching of package name will help with those call stacks too.

**Comment by r30shah** [comment]
> As an additional change, it probably makes sense to intern() the packageName String so there is only 1 copy in memory rather than 1 per class from the package.

@DanHeidinga  I was looking into it while creating change, and found that the packageName string was created and `intern()` in [1] when we compute the non-ArrayClass package name which is what we call it the class is not primitive.  

[1]. https://github.com/eclipse-openj9/openj9/blob/69ff1a02db3b0ca40fa8ed2db2cbfdc6be29afb4/jcl/src/java.base/share/classes/java/lang/Class.java#L2116-L2123

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 2]
```diff
@@ -2131,26 +2134,35 @@ private static String getNonArrayClassPackageName(Class<?> clz) {
...
-		while (componentType.isArray()) {
-			componentType = componentType.getComponentType();
+		if (isPrimitive()) {
+			packageName = "java.lang"; //$NON-NLS-1$
 		}
-		if (componentType.isPrimitive()) {
-			return "java.lang"; //$NON-NLS-1$
-		} else {
-			return getNonArrayClassPackageName(componentType);
+		else if (isArray()) {
```
Comment (tajila) [inline jcl/src/java.base/share/classes/java/lang/Class.java:2146]: nitpick: `} else if (isArray()) {`

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 3]
```diff
@@ -2131,26 +2134,35 @@ private static String getNonArrayClassPackageName(Class<?> clz) {
...
+				componentType = componentType.getComponentType();
+			}
+			if (componentType.isPrimitive()) {
+				packageName = "java.lang"; //$NON-NLS-1$
+			} else {
+				packageName = getNonArrayClassPackageName(componentType);
+			}
 		}
-	}
+		else {
```
Comment (tajila) [inline jcl/src/java.base/share/classes/java/lang/Class.java:2157]: same here

**Review by tajila — APPROVED** [review]
some minor things

**Comment by tajila** [comment]
@DanHeidinga any further comments?

**Review by DanHeidinga — APPROVED** [review]
lgtm

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 2 above)
Comment (r30shah) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: Changed in https://github.com/eclipse-openj9/openj9/compare/c1fc6e25e9028ddb5a58776f35ee9d598f2d46a2..55dd62c6660f6b9165f0b3edf80e54ea799de24c

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 3 above)
Comment (r30shah) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: Changed in https://github.com/eclipse-openj9/openj9/compare/c1fc6e25e9028ddb5a58776f35ee9d598f2d46a2..55dd62c6660f6b9165f0b3edf80e54ea799de24c

**Comment by tajila** [comment]
Jenkins test sanity zlinux jdk8,jdk17

**Pull request merged**